File: packages/go-sdk/composition/paginate.go
package composition

import "iter"

// PaginatedQuery represents pagination query parameters.
type PaginatedQuery struct {
	Cursor string
	Limit  int
}

// PaginatedResponse represents a paginated API response.
type PaginatedResponse[T any] struct {
	Data       []T    `json:"data,omitempty"`
	Items      []T    `json:"items,omitempty"`
	NextCursor string `json:"next_cursor,omitempty"`
	HasMore    *bool  `json:"has_more,omitempty"`
}

// PaginateOptions configures pagination behavior.
type PaginateOptions struct {
	Limit int
}

// Paginate returns an iterator that automatically paginates through list API responses.
func Paginate[T any](
	listFn func(query PaginatedQuery) (PaginatedResponse[T], error),
	opts *PaginateOptions,
) iter.Seq2[T, error] {
	return func(yield func(T, error) bool) {
		var cursor string
		limit := 0
		if opts != nil {
			limit = opts.Limit
		}

		for {
			q := PaginatedQuery{Cursor: cursor, Limit: limit}
			resp, err := listFn(q)
			if err != nil {
				var zero T
				yield(zero, err)
				return
			}

			items := resp.Data
			if len(items) == 0 {
				items = resp.Items
			}

			for _, item := range items {
				if !yield(item, nil) {
					return
				}
			}

			if resp.NextCursor == "" || (resp.HasMore != nil && !*resp.HasMore) || len(items) == 0 {
				return
			}

			cursor = resp.NextCursor
		}
	}
}

// CollectAll collects all items from a paginated iterator into a slice.
func CollectAll[T any](seq iter.Seq2[T, error]) ([]T, error) {
	var result []T
	for item, err := range seq {
		if err != nil {
			return nil, err
		}
		result = append(result, item)
	}
	return result, nil
}

File: packages/typescript-sdk/src/composition/paginate.ts
type PaginatedQuery = {
  readonly cursor?: string;
  readonly limit?: number;
};

type PaginatedResponse<TItem> = {
  readonly data?: readonly TItem[];
  readonly items?: readonly TItem[];
  readonly next_cursor?: string;
  readonly nextCursor?: string;
  readonly has_more?: boolean;
  readonly hasMore?: boolean;
};

/**
 * Creates an async iterator that automatically paginates through list API responses.
 *
 * Supports responses with `data` or `items` arrays, and `next_cursor`/`nextCursor`
 * for cursor-based pagination.
 *
 * @param listFn - Function that accepts a query object and returns a paginated response.
 * @param options - Optional pagination configuration.
 * @returns An async generator yielding individual items across all pages.
 *
 * @example
 * ```ts
 * // Iterate through all runs
 * for await (const run of paginate((q) => client.listRuns({ query: q }))) {
 *   console.log(run.id, run.status);
 * }
 *
 * // Collect all into an array
 * const allJobs = await collectAll(paginate((q) => client.listJobs({ query: q })));
 * ```
 */
export async function* paginate<TItem>(
  listFn: (query: PaginatedQuery) => Promise<PaginatedResponse<TItem>>,
  options?: { readonly limit?: number }
): AsyncGenerator<TItem> {
  let cursor: string | undefined;

  for (;;) {
    const response = await listFn({
      cursor,
      ...(options?.limit === undefined ? {} : { limit: options.limit }),
    });

    const items = response.data ?? response.items ?? [];

    for (const item of items) {
      yield item;
    }

    const nextCursor = response.next_cursor ?? response.nextCursor;
    const hasMore = response.has_more ?? response.hasMore;

    if (!nextCursor || hasMore === false || items.length === 0) {
      break;
    }

    cursor = nextCursor;
  }
}

/**
 * Collects all items from an async generator into an array.
 *
 * @param gen - An async generator (e.g. from {@link paginate}).
 * @returns A promise that resolves to an array of all yielded items.
 *
 * @example
 * ```ts
 * const allRuns = await collectAll(paginate((q) => client.listRuns({ query: q })));
 * ```
 */
export const collectAll = async <T>(gen: AsyncGenerator<T>): Promise<T[]> => {
  const result: T[] = [];
  for await (const item of gen) {
    result.push(item);
  }
  return result;
};
